site/content/blog/2019-06-04-percona-live-2019.md

```diff
@@ -0,0 +1,80 @@
+---
+title: Percona Live 2019 & Database Replication
+slug: percona-live-2019-database-replication
+date: 2019-06-04T12:00:00-04:00
+author: Jesse Hallam
+github: lieut-data
+community: jesse.hallam
+---
+
+This past week, I was privileged to attend [Percona Live 2019](https://www.percona.com/live/19/), an open-source database conference held this year in Austin, Texas. With the breadth of the sessions I attended being captured on a [community.mattermost.com post](https://community.mattermost.com/core/pl/ssy76opykjrd3fj35mh5sdjxuy), I wanted to use this blog post to dive into a recurring topic from the conference: database replication.
+
+#### What is replication?
+
+Replication is a means by which data on one database server is automatically copied -- or replicated -- to other database servers called replicas. Having extra copies of a database reduces the chance of data-loss in the event of a server failure. Plus, the copies can be used as databases themselves, spreading out load to improve application performance.
+
+#### How does replication work?
+
+Most types of replication fall into one of the following categories:
+
+##### Asynchronous Replication
+
+Asynchronous replication allows copied data to be out-of-date relative to the master copy. Consider a typical database interaction on a single server:
+
+![image](/blog/2019-06-04-percona-live-2019/transaction-1.svg)
+
+The client sends a `COMMIT` to confirm the transaction, but must wait for the database to flush that transaction to disk to be confident it was persisted. Otherwise, a sudden power loss after the `COMMIT` but before the `fsync()` would result in data-loss.
+
+When a replica is configured, the server receiving the write is also trying to make a copy of the data. Asynchronous replication releases the client before the copy has been confirmed on the replica:
+
+![image](/blog/2019-06-04-percona-live-2019/transaction-2.svg)
+
+A power loss before the copy is made would result in the replica being out-of-date, but it can easily catch up from the master copy on restart. Even without a power loss, there is a period of time in which the replica will be out-of-date before it receives the copy. This period of time (sometimes measured as a number of bytes) is called the replication lag.
+
+When an application directly queries an asynchronous replica, it must be prepared to deal with some replication lag. For example, such an application should never `INSERT` a row on one server and then immediately query it from another: the data won't be there yet! A healthy database keeps replication lag to a minimum. If a replica fails to keep up, it might be necessary for the application to stop querying it to avoid unexpected results.
```

Comment:
What about adding a short tidbit about how we ran into this at Mattermost and solved it? I remember we had an issue with this on user and channel member creation before

Reply:
Great idea! @jespino conveniently fixed just such an issue in v5.12.